luomu-common/src/macaddr.rs

```diff
@@ -1,61 +1,94 @@
-use std::convert::TryFrom;
+use std::convert::{TryFrom, TryInto};
 use std::fmt;
 use std::str::FromStr;
 
+use crate::TagError;
+
 use super::InvalidAddress;
 
 /// A Mac address used for example with Ethernet.
 ///
 /// Mac address is handled as big endian value. All `From<T>` implementations
-/// returning `MacAddr` expect input as big endian. `From<u64>` also expects
+/// returning [MacAddr] expect input as big endian. `From<u64>` also expects
 /// address to reside in lowest 6 bytes. All `From<MacAddr>` and
 /// `TryFrom<MacAddr>` implementations return their bytes as big endian.
+///
+/// The [MacAddr] type has space to store one VLAN ID (a tag). See
+/// [crate::TaggedMacAddr] for type with more space for storing tag stack.
 #[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
-pub struct MacAddr([u8; 6]);
+pub struct MacAddr(u64);
+
+impl MacAddr {
+    /// A unspecified MAC address (00:00:00:00:00:00)
+    pub const UNSPECIFIED: MacAddr = MacAddr(0);
+
+    /// A broadcast MAC address (FF:FF:FF:FF:FF:FF)
+    pub const BROADCAST: MacAddr = MacAddr(0x0000FFFFFFFFFFFF);
+
+    /// Return MAC address as bytearray in big endian order.
+    pub fn as_array(&self) -> [u8; 6] {
+        u64::from(self).to_be_bytes()[2..=7].try_into().unwrap()
+    }
+}
 
 impl MacAddr {
     /// Checks if this address is unspecified (`00:00:00:00:00:00`) address.
-    pub fn is_unspecified(&self) -> bool {
-        self.0 == [0; 6]
+    pub const fn is_unspecified(&self) -> bool {
+        self.0 & 0x0000FFFFFFFFFFFF == 0
     }
 
-    /// Checks if this address is multicast address.
+    /// Checks if this address is a multicast address.
     pub const fn is_multicast(&self) -> bool {
         // https://en.wikipedia.org/wiki/MAC_address#Unicast_vs._multicast_(I/G_bit)
         // The least significant bit of an address's first octet is referred to
         // as the I/G, or Individual/Group, bit. When this bit is 0 (zero),
         // the frame is meant to reach only one receiving NIC
-        (self.0[0] & 0x01) == 0x01
+        ((self.0 >> 40) & 0x01) == 0x01
     }
-}
 
-impl From<[u8; 6]> for MacAddr {
-    fn from(val: [u8; 6]) -> Self {
-        Self(val)
+    /// Push VLAN tag.
+    pub const fn push_tag(&mut self, tag: u16) -> Result<(), TagError> {
+        if tag > 0x0FFF {
+            return Err(TagError::TooLargeTag);
+        }
+
+        self.0 = (self.0 & 0xF000FFFFFFFFFFFF) | ((tag as u64) << 48);
+        Ok(())
     }
-}
 
-impl From<&[u8; 6]> for MacAddr {
-    fn from(val: &[u8; 6]) -> Self {
-        Self(val.to_owned())
+    /// Pop a VLAN tag.
+    pub const fn pop_tag(&mut self) -> Option<u16> {
+        let Some(tag) = self.peek_tag() else {
```

Comment:
This could be written as 
```
let tag = self.peek_tag()?;
```

Reply:
`?` is not allowed in a `const fn`. See issue #74935 <https://github.com/rust-lang/rust/issues/74935> for more information.